### PR TITLE
dnn: refactor reduce

### PR DESCRIPTION
**Pull request by fengyuentau**
Reasons of refactoring reduce layer:
- Previous impl is not a single operation impl, leading to difficulties for other backend to support reduce layer.
- [need to validate] Previous impl does not consider a case when reduced axes are symmetric.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by zihaomu — APPROVED** [review]
LGTM! 👍

**Comment by fengyuentau** [comment]
Some variables are removed as they are not needed with this patch, causing [some alerts in default](https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/103396/steps/Compare%20ABI%20dumps/logs/report-html#Removed), but it should be fine.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 // Third party copyrights are property of their respective owners.
 
 #include "../precomp.hpp"
+#include "opencv2/core/check.hpp"
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_importer.cpp:9]: Thank you for the notice. Clangd defaults on inserting headers ...

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -911,7 +912,7 @@ void ONNXImporter::populateNet()
     for(int li = 0; li < layersSize; li++)
     {
         const opencv_onnx::NodeProto& node_proto = graph_proto.node(li);
-        handleNode(node_proto);
+         handleNode(node_proto);
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_importer.cpp]: fixed.

**Review by vpisarev — APPROVED** [review]

**Pull request merged**